pacsman/dcmtk_client.py

```diff
@@ -0,0 +1,423 @@
+        set_undefined_tags_to_blank(search_dataset, additional_tags)
+
+        id_responses = self._send_c_find(search_dataset)
+        for study in id_responses:
+            if hasattr(study, 'PatientID'):
+                result = patient_id_to_datasets[study.PatientID]
+                self.update_patient_result(result, study, additional_tags)
+
+        # then search with the same query on the patient name field
+        search_dataset = self._get_study_search_dataset()
+        search_dataset.PatientName = search_query
+        set_undefined_tags_to_blank(search_dataset, additional_tags)
+
+        name_responses = self._send_c_find(search_dataset)
+        for study in name_responses:
+            if hasattr(study, 'PatientID'):
+                result = patient_id_to_datasets[study.PatientID]
+                self.update_patient_result(result, study, additional_tags)
+
+        return list(patient_id_to_datasets.values())
+
+    def studies_for_patient(self, patient_id, additional_tags=None) -> List[Dataset]:
+        search_dataset = self._get_study_search_dataset()
+        search_dataset.PatientID = patient_id
+        set_undefined_tags_to_blank(search_dataset, additional_tags)
+
+        responses = self._send_c_find(search_dataset)
+        datasets = []
+        for dataset in responses:
+            # Some PACS send back empty "Success" responses at the end of the list
+            if hasattr(dataset, 'PatientID'):
+                datasets.append(dataset)
+
+        return datasets
+
+    def search_series(self, query_dataset, additional_tags=None) -> List[Dataset]:
+        additional_tags = additional_tags or []
+        query_dataset.QueryRetrieveLevel = 'IMAGE'
+        additional_tags += [
+            'Modality',
+            'BodyPartExamined',
+            'SeriesDescription',
+            'SeriesDate',
+            'SeriesTime',
+            'PatientPosition',
+        ]
+        set_undefined_tags_to_blank(query_dataset, additional_tags)
+
+        datasets = []
+        responses = self._send_c_find(query_dataset)
+        for series in responses:
+            if hasattr(series, 'SeriesInstanceUID'):
+                datasets.append(series)
+        return datasets
+
+    def series_for_study(self, study_id, modality_filter=None, additional_tags=None) -> \
+            List[Dataset]:
+        additional_tags = additional_tags or []
+
+        dataset = Dataset()
+        dataset.StudyInstanceUID = study_id
+        dataset.QueryRetrieveLevel = 'SERIES'
+
+        additional_tags += [
+            'SeriesInstanceUID',
+            'BodyPartExamined',
+            'SeriesDescription',
+            'SeriesDate',
+            'SeriesTime',
+            'StudyDate',
+            'StudyTime',
+            'PatientPosition',
+            'NumberOfSeriesRelatedInstances',
+        ]
+        set_undefined_tags_to_blank(dataset, additional_tags)
+        # TODO check if filtering modality with 'MR\\CT' works in dcmtk
```

Comment:
IIRC it's going to need to be multiple queries

Reply:
I'll just take this TODO off then. It may just be easier to filter outside of pacsman if more queries are required.